Explicit help message for binop type mismatch

estebank opened the pull request:
When trying to do `1 + Some(2)`, or some other binary operation on two
types different types without an appropriate trait implementation, provide
an explicit help message:

```rust
help: `{integer} + std::option::Option<{integer}>` has no implementation
```

Re: #39579, #38564, #37626, #39942, #34698.

rust-highfive commented:
r? @pnkfelix

(rust_highfive has picked a reviewer for you, use r? to override)

estebank commented:
[WIP] Missing tests ATM.

estebank commented:
CCing the people involved in the referenced tickets: @beamspease, @abonander, @brson, @killercup, @oli-obk, @frewsxcv 

Could I get some feedback on the current output of this PR?

<img width="859" alt="" src="https://cloud.githubusercontent.com/assets/1606434/23981628/3f18c738-09c5-11e7-8a5e-982104da1021.png">

abonander commented:
I like it. 

This isn't specific to this PR, I know, but I kind of wish that error messages didn't use full paths for types, because that's just noise. I would like to see error messages use type names based on how they're imported, but resolving that is nontrivial.

estebank commented:
> What does it print for `PartialOrd` operators?

@abonander, at the moment it's keeping [the current output](https://is.gd/bd1IE4) for [`PartialOrd`, `<<` and `>>`](https://github.com/rust-lang/rust/pull/40565/files#diff-0e12890ad498e250783f40f88a8b8ec6R571).

oli-obk commented on src/librustc/traits/error_reporting.rs:
```diff
@@ -550,10 +550,33 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+                                                  op,
+                                                  trait_ref.substs()[1]));
+                            } else if &trait_def == "std::cmp::PartialEq" {
+                                err.help(&format!("can't compare `{}` against `{:?}`",
```
Both could be Display format specifiers

estebank replied:
`Kind` doesn't implement `Display` and using `Ty` instead would not work for [cases with `impl`](https://github.com/rust-lang/rust/pull/40565/files#diff-34dc3d265b15de4ce7dbb7e9b848d42cR36).

oli-obk replied:
"with" instead of "against"

estebank replied:
Done.

oli-obk commented on src/librustc/traits/error_reporting.rs:
```diff
@@ -550,10 +550,33 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
-                            // similar impls.
-                            let impl_candidates = self.find_similar_impl_candidates(trait_ref);
-                            self.report_similar_impl_candidates(impl_candidates, &mut err);
+                            let binops: FxHashMap<&str, &str> = [
```
This won't work on no_std.

estebank replied:
Could you clarify what wouldn't work?

oli-obk replied:
In a crate with no_std, these path start with `core`

estebank replied:
Added `core::*` versions to the list. Feels hacky...

arielb1 replied:
Use lang items please - `tcx.lang_items.add_trait` etc. - instead of guessing stuff.

estebank replied:
Done.

oli-obk commented on src/test/ui/span/multiline-span-simple.stderr:
```diff
@@ -9,11 +9,7 @@ error[E0277]: the trait bound `u32: std::ops::Add<()>` is not satisfied
-             <&'a u32 as std::ops::Add<u32>>
-             <u32 as std::ops::Add<&'a u32>>
-             <&'b u32 as std::ops::Add<&'a u32>>
+   = help: `u32 + ()` has no implementation
```
The suggestions should still show up. Maybe even in the Format oft this PR?

estebank replied:
I felt that for these cases in particular, in order to keep it newbie friendly, it was better to simplify the text in order to make the error less daunting.

oli-obk replied:
In case of `&a + a` I think it would be very helpful to see the hints

estebank replied:
Fair enough. Should I _only_ show them in the `&a + a` case then?

oli-obk replied:
Would be enough for the cases I've run into with ppl I've taught rust

oli-obk commented on src/test/ui/mismatched_types/binops.stderr:
```diff
@@ -0,0 +1,26 @@
+13 |     1 as usize - Some(1);
+   |     ^^^^^^^^^^^^^^^^^^^^ the trait `std::ops::Sub<std::option::Option<{integer}>>` is not implemented for `usize`
+   |
+   = help: `usize - std::option::Option<{integer}>` has no implementation
```
The note is already duplicating the main error message. You could remove the note and make your help a note

estebank replied:
It makes sense. Something along the lines of this?:

```rust
error[E0277]: the trait bound `usize: std::ops::Sub<std::option::Option<{integer}>>` is not satisfied
  --> $DIR/binops.rs:13:5
   |
13 |     1 as usize - Some(1);
   |     ^^^^^^^^^^^^^^^^^^^^ no implementation for `usize - std::option::Option<{integer}>`
```

oli-obk replied:
Yes

estebank replied:
Done.

nagisa commented:
Message like this can be introduced without changing the compiler by adding a #[on_unimplemented] attribute onto the traits.

estebank commented:
> Message like this can be introduced without changing the compiler by adding a `#[on_unimplemented]` attribute onto the traits.

That wouldn't let me [remove the "alternative implementations" recomendation](https://github.com/rust-lang/rust/pull/40565/files#diff-1df5a9ddcdb3d4e6220509d4e573b810L12) or [replace the span label](https://github.com/rust-lang/rust/pull/40565/files#diff-ca519adaaa250d1a0b7d7339001704c9R5).

estebank commented:
Current output:

![](https://cloud.githubusercontent.com/assets/1606434/24181152/657726ee-0e76-11e7-8d9e-3fa638b2c4d0.png)

beamspease commented:
Very nice

estebank commented:
Ping.

CC @jonathandturner, @GuillaumeGomez, @nikomatsakis

jonathandturner commented:
Oh I'm glad you're tackling this one!

Can you add tests for a couple more operators?  I like the way it looks, but just wanted to make sure other operators looked like it seems like they should.

GuillaumeGomez commented:
Awesome! :D

arielb1 commented:
@estebank 

What's the problem with using `#[rustc_on_unimplemented]` again? That it does not affect the primary label? Maybe it should instead of being a note?

estebank commented:
I really want to remove the suggestions because for these cases it is more confusing than helpful:

```rust
error[E0277]: the trait bound `{integer}: std::ops::Add<std::option::Option<{integer}>>` is not satisfied
 --> <anon>:3:5
  |
3 |     1 + Some(1);
  |     ^^^^^^^^^^^ the trait `std::ops::Add<std::option::Option<{integer}>>` is not implemented for `{integer}`
  |
  = help: the following implementations were found:
            <std::borrow::Cow<'a, str> as std::ops::Add<&'a str>>
            <std::borrow::Cow<'a, str> as std::ops::Add>
            <u32 as std::ops::Add>
            <&'a u32 as std::ops::Add<u32>>
          and 106 others

error[E0277]: the trait bound `usize: std::ops::Sub<std::option::Option<{integer}>>` is not satisfied
 --> <anon>:4:5
  |
4 |     1 as usize - Some(1);
  |     ^^^^^^^^^^^^^^^^^^^^ the trait `std::ops::Sub<std::option::Option<{integer}>>` is not implemented for `usize`
  |
  = help: the following implementations were found:
            <usize as std::ops::Sub>
            <&'a usize as std::ops::Sub<usize>>
            <usize as std::ops::Sub<&'a usize>>
            <&'b usize as std::ops::Sub<&'a usize>>

error[E0277]: the trait bound `{integer}: std::cmp::PartialEq<std::result::Result<{integer}, _>>` is not satisfied
 --> <anon>:5:5
  |
5 |     1 == Ok(1);
  |     ^^^^^^^^^^ the trait `std::cmp::PartialEq<std::result::Result<{integer}, _>>` is not implemented for `{integer}`
  |
  = help: the following implementations were found:
            <&'a A as std::cmp::PartialEq<&'b B>>
            <&'a mut A as std::cmp::PartialEq<&'b mut B>>
            <&'a A as std::cmp::PartialEq<&'b mut B>>
            <&'a mut A as std::cmp::PartialEq<&'b B>>
          and 697 others
```

vs

```rust

error[E0277]: the trait bound `{integer}: std::ops::Add<std::option::Option<{integer}>>` is not satisfied
  --> $DIR/binops.rs:12:5
   |
12 |     1 + Some(1);
   |     ^^^^^^^^^^^ no implementation for `{integer} + std::option::Option<{integer}>`

error[E0277]: the trait bound `usize: std::ops::Sub<std::option::Option<{integer}>>` is not satisfied
  --> $DIR/binops.rs:13:5
   |
13 |     1 as usize - Some(1);
   |     ^^^^^^^^^^^^^^^^^^^^ no implementation for `usize - std::option::Option<{integer}>`

error[E0277]: the trait bound `{integer}: std::cmp::PartialEq<std::result::Result<{integer}, _>>` is not satisfied
  --> $DIR/binops.rs:14:5
   |
14 |     1 == Ok(1);
   |     ^^^^^^^^^^ can't compare `{integer}` with `std::result::Result<{integer}, _>`
```

Using `rustc_on_unimplemented` the output would look this way:

estebank commented:
You're right, of course. I don't think there's a way to get the types involved on `rustc_on_unimplemented`, though, to accomplish the same kind of message proposed on this PR.

arielb1 commented:
@estebank

What's the problem with adding something like:
```Rust
#[rustc_on_unimplemented = "no implementation for `{Self} + {RHS}`"]
```

estebank commented:
@arielb1 I'm so sorry, I only now see what you mean. I had completely missed that [this attribute takes arguments](https://doc.rust-lang.org/reference.html#miscellaneous-attributes). Then, the difference between this PR and your suggestion is much smaller than I thought:

```rust
error[E0277]: the trait bound `{integer}: std::ops::Add<std::option::Option<{integer}>>` is not satisfied
  --> $DIR/binops.rs:12:5
   |
12 |     1 + Some(1);
   |     ^^^^^^^^^^^ no implementation for `{integer} + std::option::Option<{integer}>`

error[E0277]: the trait bound `usize: std::ops::Sub<std::option::Option<{integer}>>` is not satisfied
  --> $DIR/binops.rs:13:5
   |
13 |     1 as usize - Some(1);
   |     ^^^^^^^^^^^^^^^^^^^^ no implementation for `usize - std::option::Option<{integer}>`

error[E0277]: the trait bound `{integer}: std::cmp::PartialEq<std::result::Result<{integer}, _>>` is not satisfied
  --> $DIR/binops.rs:14:5
   |
14 |     1 == Ok(1);
   |     ^^^^^^^^^^ can't compare `{integer}` with `std::result::Result<{integer}, _>`
```

vs

```rust
error[E0277]: the trait bound `{integer}: std::ops::Add<std::option::Option<{integer}>>` is not satisfied
 --> <anon>:3:5
  |
3 |     1 + Some(1);
  |     ^^^^^^^^^^^ the trait `std::ops::Add<std::option::Option<{integer}>>` is not implemented for `{integer}`
  |
  = note: no implementation for `{integer} + std::option::Option<{integer}>`

error[E0277]: the trait bound `usize: std::ops::Sub<std::option::Option<{integer}>>` is not satisfied
 --> <anon>:4:5
  |
4 |     1 as usize - Some(1);
  |     ^^^^^^^^^^^^^^^^^^^^ the trait `std::ops::Sub<std::option::Option<{integer}>>` is not implemented for `usize`
  |
  = note: no implementation for `usize + std::option::Option<{integer}>`

error[E0277]: the trait bound `{integer}: std::cmp::PartialEq<std::result::Result<{integer}, _>>` is not satisfied
 --> <anon>:5:5
  |
5 |     1 == Ok(1);
  |     ^^^^^^^^^^ the trait `std::cmp::PartialEq<std::result::Result<{integer}, _>>` is not implemented for `{integer}`
  |
  = note: can't compare `{integer}` with `std::result::Result<{integer}, _>`
```

I still believe the first version is nicer for newcomers, but the second option is already a huge improvement over the status quo. I'll be updating this PR to use `rustc_on_unimplemented` instead. Sorry for the misunderstanding on my part.

arielb1 commented on src/librustc/traits/error_reporting.rs:
```diff
@@ -487,15 +487,15 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                                   error: &SelectionError<'tcx>)
     {
-        let span = obligation.cause.span;
+        let sp = obligation.cause.span;
```
Could you change this back to `span` and squash?

estebank replied:
Done

arielb1 commented on src/librustc/traits/error_reporting.rs:
```diff
@@ -544,25 +537,28 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
+                        } else if let Some(s) = self.on_unimplemented_note(trait_ref, obligation) {
                             // If it has a custom "#[rustc_on_unimplemented]"
                             // error message, let's display it!
                             err.note(&s);
```
We can change all `on_unimplemented` messages to be labels by making this a `.span_label` call. Not sure it's a good idea.

estebank replied:
Probably a bad idea to do that for all cases, but I've been considering the possibility of having a way to configure wether it is shown as a note or as the span's label. I'm thinking that it could either accomplished by changing the parsing of attributes to accept extra arguments, or simply have it change behavior when the `rustc_on_unimplemented` text starts with `label: `. It might be too more magic than I'm comfortable with, though.

arielb1 commented:
r+ modulo squash.

estebank committed:
Explicit help message for binop type missmatch

When trying to do a binary operation with missing implementation, for
example `1 + Some(2)`, provide an explicit help message:

```
note: no implementation for `{integer} + std::option::Option<{integer}>`
```

Use `rustc_on_unimplemented` for the suggestions. Move cfail test to ui.

estebank commented:
@bors r=arielb1

@arielb1 squashed into single commit.

bors commented:
:pushpin: Commit 9aa75a8 has been approved by `arielb1`

<!-- @bors r=arielb1 9aa75a881a178c64ea25f453deda3a88081cd7c2 -->

estebank commented:
@bors r-

arielb1 commented:
@bors r+

bors commented:
:pushpin: Commit 9aa75a8 has been approved by `arielb1`

<!-- @bors r=arielb1 9aa75a881a178c64ea25f453deda3a88081cd7c2 -->

estebank commented:
@bors r=arielb1

hadn't seen the comment about `sp` -> `span`, changed.

bors commented:
:pushpin: Commit be8787d has been approved by `arielb1`

<!-- @bors r=arielb1 be8787dfe564cf315a9343a84724130da322e805 -->

bors commented:
:hourglass: Testing commit be8787d with merge 730e5ad...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.2870), [status-travis](https://travis-ci.org/rust-lang/rust/builds/220740848)
Approved by: arielb1
Pushing 730e5ad04e23f30cc24e4b87dfd5da807325e243 to master...
